### ISSUE TITLE
Issue with gateway post-processing for @interfaceObject results

### ISSUE THREAD

**Issue by pcmanus** [apollographql/federation#2320]
There is currently an issue when using `@interfaceObject` and data is fetched without resolving the concrete type. That is, to illustrate, if one has:
```graphql
type Query {
  t1: T
}

interface T @key(fields: "id") {
  id: ID!
}

# ... some implementations of `T`, details don't matter
```

```graphql
type Query {
  t2: T
}

type T @key(fields: "id") @interfaceObject {
  id: ID!
  v: Int
}
```
and the query is:
```graphql
query {
  t2 {
    v
  }
}
```
Then the intent is that we only need to query the 2nd subgraph and never need to resolve the underlying implementation type, since only fields of the interface is requested and `__typename` is not explicitly requested. And indeed, the query plan will simply send the query essentially as-is to the 2nd subgraph.

But while this correctly fetch the proper results internally, the gateway run the graphQL execution on those internal results as way to filter/reorder the subset of the internal results than needs to be returned, and this currently does not work in such example because in the supergraph, type `T` is an interface, and so graphQL-js execute _requires_ that that we have a `__typename` with a real implementation name, which we don't have. The result is that the code currently gets nullified instead of returning the proper results.

Overall, this is a restriction of our post-processing use of graphQL-js execute: we don't really want a full execution and we have the data for the field queried and the implementation type is not requested in any way, so we ought to be able to return the results.

The "right" fix here is imo to change our post-processing, and to stop using the full graphQL-js `execute` and instead add code that just do the bits of filtering and reordering on the internal data that we need, at which point that code can easily handle this case.

I'll note that this is something that has come up before as calling graphQL-js `execute` is known to have a somewhat large performance cost, despite us not needing a lot of what is done. There is even [a longstanding FIXME](https://github.com/apollographql/federation/blob/main/gateway-js/src/executeQueryPlan.ts#L96) in the code taht says that this should be improved/simplified.

Tl;dr, I suggest that it's time that we fix that FIXME once and for all.
